lib/train/transports/clients/azure/vault.rb

```diff
@@ -0,0 +1,36 @@
+# encoding: utf-8
+# frozen_string_literal: true
+
+require 'azure_mgmt_key_vault'
+
+# Wrapper class for ::Azure::KeyVault::Profiles::Latest::Mgmt::Client allowing custom configuration,
+# for example, defining additional settings for the ::MsRestAzure::ApplicationTokenProvider.
+class Vault
+  def initialize(vault_name)
+    if vault_name.nil?
+      raise ::Train::UserError, 'Vault Name cannot be nil'
+    end
+    @auth_endpoint = MsRestAzure::AzureEnvironments::AzureCloud.active_directory_endpoint_url
+    @api_endpoint = "https://#{vault_name}#{MsRestAzure::AzureEnvironments::AzureCloud.key_vault_dns_suffix}"
+    @resource_endpoint = 'https://vault.azure.net'
+  end
```

Comment:
Ah, was `GraphRbac` changed to make it match this API? If so, I don't think this should be stateful either. Both of these classes take some input and return the appropriate client, and are then discarded. No further input or other reason for change over time.

```ruby
class Vault
  AUTH_ENDPOINT     = MsRestAzure::AzureEnvironments::AzureCloud.active_directory_endpoint_url
  RESOURCE_ENDPOINT = 'https://vault.azure.net'

  def self.client(vault_name, credentials)
    raise ::Train::UserError, 'Vault Name cannot be nil' unless vault_name

    ::Azure::KeyVault::Profiles::Latest::Mgmt::Client.new(
      credentials.merge(
        credentials: ::MsRest::TokenCredentials.new(provider(credentials, settings)),
￼        base_url: api_endpoint(vault_name)
      )
    )
  end

  def self.api_endpoint(vault_name)
    "https://#{vault_name}#{MsRestAzure::AzureEnvironments::AzureCloud.key_vault_dns_suffix}"
  end

  def self.provider(credentials, settings)
    ::MsRestAzure::ApplicationTokenProvider.new(
￼       credentials[:tenant_id],
￼       credentials[:client_id],
￼       credentials[:client_secret],
￼       settings,
￼     )
  end

  def self.settings
￼     client_settings = ::MsRestAzure::ActiveDirectoryServiceSettings.get_azure_settings
￼     client_settings.authentication_endpoint = AUTH_ENDPOINT
￼     client_settings.token_audience = RESOURCE_ENDPOINT
￼     client_settings
￼   end
end
```

You could just make it a module then, of course, and replace the `self.` prefixes with entries in `module_function`, making the internal implementation methods private with `private_class_method`. But that's all a bit out of step with the conventions in this library.
```ruby
module Vault
  def client(vault_name, credentials)
  # ...

  def api_endpoint(vault_name)
  # ...

  def provider(credentials, settings)
  # ...

  def settings
  # ...

  module_function :client, :api_endpoint, :provider, :settings
  private_class_method :api_endpoint, :provider, :settings
end
```

Reply:
Thanks for this! I went for something very similar - will ping you about it.